### ISSUE TITLE
Ltac handles open terms incorrectly in `constr`

### ISSUE THREAD

**Issue by JasonGross** [coq/coq#7195]
#### Version

8.8+beta1

#### Description of the problem

This code should work.  It does not.
```coq
Goal True.
  let x := constr:((O, tt)) in
  let rx := lazymatch (eval pattern nat, unit, O, tt in x) with
            | ?rx _ _ _ _ => rx end in
  idtac rx (* (fun (P P0 : Set) (n : P) (u : P0) => (n, u)) *);
    let rx := lazymatch rx with fun (N : Set) => ?rx => constr:(fun N : Type => rx) end in
    idtac rx (* (fun (P : Type) (P0 : Set) (n : P) (u : P0) => (n, u)) *);
      let __ := type of rx in
      unify x (rx nat unit O tt);
      let rx := lazymatch rx with fun (N : Type) (U : Set) => ?rx => constr:(fun (N : Type) (U : Type) => rx) end in
      idtac rx (* (fun (_ P0 : Type) (n u : P0) => (n, u)) *);
      let __ := type of rx in
      unify x (rx nat unit O tt).
  (* Error: The term "0" has type "nat" while it is expected to have type "unit". *)
```

**Comment by SkySkimmer** [comment]
I'm guessing the issue is that the P0 is just a print-time renamed P so 
when you match N and U are both P.
Someone who knows ltac matching tell me if I'm right.

Gaëtan Gilbert

**Comment by JasonGross** [comment]
Yes, I'm pretty sure that's correct.  Here's another example:
```coq
Goal True.
  let x := (eval cbv beta zeta in (fun P => let Q := P in fun P => P + Q)) in
  unify x (fun a b => b + a); (* success *)
    let x' := lazymatch x with
              | (fun (a : ?A) (b : ?B) => ?k)
                => constr:(fun (a : A) (b : B) => k)
              end in
    unify x x'.
  (* Error:
Ltac call to "unify (constr) (constr)" failed.
Tactic failure: Not unifiable. *)
```

**Comment by herbelin** [comment]
Fix at #7199 (note that this fails since 8.5).
